### DISCUSSION TITLE
Questions regarding this project

### DISCUSSION THREAD

**Discussion by Joe23232** [emilk/egui#284] [General]
Hi I have some questions regarding this project:

1. Is it possible to build Android applications by using this crate as well as the [`android-ndk-rs`](https://github.com/rust-windowing/android-ndk-rs)?
2. How does this project differ to [`egui`](https://github.com/emilk/egui)
3. Is it in its stable form?
4. Does this have retained mode as well?

**Comment by emilk** [comment]
1. no idea
2. this is `egui`
3. [no](https://github.com/emilk/egui#state)
4. no, `egui` is [pure immediate mode](https://github.com/emilk/egui#goals)

**Reply by Joe23232** [reply]
2. Sorry I meant [fltk-rs](https://github.com/MoAlyousef/fltk-rs)
3. I see
4. Is it reactive or continuous?

**Reply by emilk** [reply]
it can be set up to be reactive. That's the default for the web demo, but you can change it in the "Backend" panel: https://emilk.github.io/egui/index.html

**Reply by Joe23232** [reply]
I see thanks

**Reply by Joe23232** [reply]
@emilk If I set it up to be reactive rather than continuous for egui, then would it still take more professing power compared to retained graphics? Would it take more power compared to retained if it was being operated on a laptop battery?

**Reply by emilk** [reply]
Maybe? egui is optimized for the case of a lot of changes on screen, while retained mode libraries are optimized for static ui:s. If you are showing live data that is changing often, maybe egui will win. If you are showing a static gui than only changes when you click a button, maybe retained will win. Maybe the difference is minimal compared to the CPU used by the rest of your app.

It really depends on the app, and what retained mode library you compare it to.

I would suggest implementing a simple GUI in both egui and e.g. iced and compare!

**Reply by Joe23232** [reply]
I see thanks mate, and with iced it is retained, right?

**Reply by Joe23232** [reply]
@emilk Hey bro one question.

Using `egui`, while I believe it is possible to create widgets inside the window, is it possible to also directly display it to the `wayland` display server without much changes to the `egui` code?

Would it be recommended to use retained graphics or immediate graphics for widgets that needs to update once per every second to show different status such as CPU usage, etc?

Am I able to design these kinds of widgets:

![Desktop](https://user-images.githubusercontent.com/34926497/134434829-0a37a7f7-8be1-4b27-8fa4-715e03563724.png)

**Comment by gravit22** [comment]
It doesn't have a backend for android. I tested it on https://github.com/rust-windowing/android-ndk-rs. 
But is it planned?

**Reply by emilk** [reply]
I have no plans for android, but it would be great if someone gave it a go!
